### ISSUE TITLE
DeleteThread to remove observers for Private Thread with more than 2 users

### ISSUE THREAD

**Issue by CalvinTp** [chat-sdk/chat-sdk-ios#9]
Hi Ben, Simon,
ChatSDK seems to have a bug in the following code; when there is a Private Thread with more than 2 users, the following methods will not be called:
- [self off];
- [self messagesOff];

It is understandable that, it is robust handling to make sure if this is a Private thread with 2 users, the observers must be maintained for thread "Details" and "Messages", however when there are 2 users, the method to remove the observers seems to be accidentally left out. 

Code pasted below as a reference: 

```
    if (_model.type.intValue & bThreadTypePrivate && _model.users.allObjects.count == 2) {
        // Rather than delete the thread we just set the status as deleted
        [currentThreadUser setValue:@{bNameKey: currentUser.name, bDeletedKey: [FIRServerValue timestamp]}
                withCompletionBlock:^(NSError * error, FIRDatabaseReference * ref) {
                    if (!error) {
                        [promise resolveWithResult:Nil];
                    }
                    else {
                        [promise rejectWithReason:error];
                    }
                }];
        
        return promise.thenOnMain(^id(id success)
                                  {
                                      [[BStorageManager sharedManager].a save];
                                      // We can keep listening to the thread. That way, if a new message comes in,
                                      // it get's regenerated
                                      //[self off];
                                      //[self messagesOff];
                                      [[NSNotificationCenter defaultCenter] postNotificationName:bNotificationThreadDeleted object:Nil];
                                      
                                      return Nil;
                                      
                                  }, ^id(NSError * error)
                                  {
                                      [[BStorageManager sharedManager].a undo];
                                      return error;
                                  });
    }
    else {
     // Otherwise we just remove the user
        return [self removeUser:[CCUserWrapper userWithModel:currentUser]];
    }

```

Thanks,
Calvin

**Comment by simonsmiley64** [comment]
Hi Calvin,

Looking through this functionality, it is intentionally set up like this. This is because the functionality is to delete and not block the user in a 1-to-1 conversation. 

For example:

1-to-1 conversation:
I am in a conversation with you and you delete the thread. If we call [self off] and [self messageOff] then you will not receive any messages from me. This is effectively blocking the user. 

**Note:** It is worth noting that when the app next returns to foreground observers will be added to all threads again.

In the current situation we delete the thread and this sets a delete status on the thread. If the user opens the thread the conversation starts at the point of delete. If you want to refresh the messages you can pull down to refresh them all back. 

If you truly don't want to receive any messages from a user then blocking them would be the best option. Otherwise you could modify the refresh functionality so that all previous messages were blocked for the user who deleted the thread.

Simon

**Comment by bensmiley** [comment]
When we have a 1-to-1 thread, we will never turn the thread off. However, when we leave a group chat, we do want to turn the thread off. 

I've added additional code to the state manager to listen to when we are removed from a Thread - when this happens, the thread is turned off.

This issue will be fixed in 4.2.1
